### PR TITLE
(docs) Update 'Using plugins' page, add 'Supported plugins' page

### PR DESCRIPTION
**Pull request by beechtom**
This updates the 'Using plugins' page to describe using plugins in more
detail. It also moves the list of bundled plugins to a new 'Supported
plugins' page.

!no-release-note

### CODE REVIEW

File: documentation/using_plugins.md [hunk 1]
```diff
@@ -1,339 +1,213 @@
...
-multiple hooks. The way in which a plugin is used varies depending on the type
-of hook used.
-
-> **Note:** Plugins are only available in configuration files and version 2
-> inventory files.
-
+Bolt supports the use of plugins to dynamically load information during a
+Bolt run and change how some Bolt actions are executed. Plugins either ship
+with Bolt or are installed as Puppet modules that have the same name as the
+plugin.
```
[the remark this answers is not in the document]
Comment (hestonhoffman) [reply documentation/using_plugins.md:6]: Actually, this makes me wonder if we need a short section here on installing a plugin from another user? Even if it's a case of linking off to the installing modules page. 
Edit: It looks like users might have to use the built-in task plugin?

File: documentation/supported_plugins.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
+# Supported plugins
+
+The following plugins are supported and maintained by Bolt. Supported plugins
+are shipped with Bolt packages and do not need to be installed separately.
+
+## Reference plugins
+
+Reference plugins are used to fetch data from an external source and store it in
+a static data object. Reference plugins can be used in configuration files,
+inventory files, and plans.
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:10]:
```suggestion
Reference plugins fetch data from an external source and store it in
a static data object. You can use reference plugins in configuration files,
inventory files, and plans.
```

File: documentation/bolt.ditamap [hunk 3]
```diff
@@ -16,9 +16,11 @@
         <topicref href="projects.md" format="markdown"/>
         <topicref href="inventory_file_v2.md" format="markdown"/>
         <topicref href="configuring_bolt.md" format="markdown"/>
-        <topicref href="using_plugins.md" format="markdown">
+        <topichead navtitle="Plugins">
```
Comment (hestonhoffman) [inline documentation/bolt.ditamap:19]: After reading through this stuff, I think we can move plugins out of the configuring section and just list them in the main nav. Their application is much broader than just configuration. Maybe this section can go after modules.

File: documentation/supported_plugins.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+| `prompt` | Prompt the user for a sensitive value. | [prompt](#prompt) |
+| `puppetdb` | Query PuppetDB for a group of targets. | [puppetdb](#puppetdb) |
+| `task` | Run a task as a plugin. | [task](#task) |
+| `terraform` | Generate targets from local and remote Terraform state files. | [terraform](https://forge.puppet.com/puppetlabs/terraform) |
+| `vault` | Access secrets from a Key/Value engine on a Hashicorp Vault server. | [vault](https://forge.puppet.com/puppetlabs/vault) |
+| `yaml` | Compose multiple YAML files into a single file. | [yaml](#yaml) |
+
+## Secret plugins
+
+Secret plugins are used to create keys for encryption and decryption, encrypt
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:28]:
```suggestion
Use secret plugins to create keys for encryption and decryption, encrypt
```

File: documentation/supported_plugins.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+plaintext, and decrypt ciphertext. Secret plugins are used by Bolt's `secret`
+command.
+
+| Plugin | Description | Documentation |
+| --- | --- | --- |
+| `pkcs7` | Generate key pairs, encrypt plaintext, and decrypt ciphertext. | [pkcs7](https://forge.puppet.com/puppetlabs/pkcs7) |
+
+## Puppet library plugins
+
+Puppet library plugins are used to make sure the Puppet library is installed
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:38]:
```suggestion
Puppet library plugins ensure that the Puppet library is installed
```

File: documentation/supported_plugins.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+on a target when the `apply_prep` function is called.
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:39]:
```suggestion
on a target when a plan calls the `apply_prep` function.
```

File: documentation/supported_plugins.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+and load them into an inventory or configuration file.
+
+#### Parameters
+
+The following parameters are available to the `env_var` plugin:
+
+| Parameter | Description | Type | Default |
+| --- | ----------- | ---- | ------- |
+| `var` | **Required.** The name of the environment variable to read from. | `String` | None |
+| `default` | A value to use if the environment variable `var` isn't set | `String` | None |
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:61]:
```suggestion
| `default` | A value to use if the environment variable `var` isn't set. | `String` | None |
```

File: documentation/supported_plugins.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+| `optional` | Unless `true`, env_var will raise an error when the environment variable `var` does not exist.  When `optional` is `true` and `var` does not exist, env_var returns `nil` | `Boolean` | `false` |
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:62]:
```suggestion
| `optional` | Unless `true`, `env_var` raises an error when the environment variable `var` does not exist.  When `optional` is `true` and `var` does not exist, `env_var` returns `nil`. | `Boolean` | `false` |
```

File: documentation/supported_plugins.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+The `puppetdb` plugin queries PuppetDB for a group of targets. 
+
+If target-specific configuration is required, the `puppetdb` plugin can be used
+to lookup configuration values for the `alias`, `config`, `facts`, `features`,
+`name`, `uri` and `vars` inventory options for each target. These values can be
+set in the `target_mapping` field. The fact lookup values can be either
+`certname` to reference the `[certname]` of the target, or a [PQL dot
+notation](https://puppet.com/docs/puppetdb/latest/api/query/v4/ast.html#dot-notation)
+facts string such as `facts.os.family` to reference a fact value. Dot notation
+is required for both structured and unstructured facts.
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:119]:
```suggestion
If you require target-specific configuration, you can use the `puppetdb` plugin
to look up configuration values for the `alias`, `config`, `facts`, `features`,
`name`, `uri` and `vars` inventory options for each target. Set these values
in the `target_mapping` field. The fact lookup values can be either
`certname` to reference the `[certname]` of the target, or a [PQL dot
notation](https://puppet.com/docs/puppetdb/latest/api/query/v4/ast.html#dot-notation)
facts string such as `facts.os.family` to reference a fact value. Dot notation
is required for both structured and unstructured facts.
```

File: documentation/supported_plugins.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+The following parameters are available to the `puppetdb` plugin:
+
+| Parameter | Description | Type | Default |
+| --- | ----------- | ---- | ------- |
+| `query` | **Required.** A string containing a [PQL query](https://puppet.com/docs/puppetdb/latest/api/query/v4/pql.html) or an array containing a [PuppetDB AST format query](https://puppet.com/docs/puppetdb/latest/api/query/v4/ast.html). | `String` | None |
+| `target_mapping` | **Required.** A hash of target attributes (`name`, `uri`, `config`) to populate with fact lookup values. | `Hash` | None |
+
+> **Note:** If neither `name` nor `uri` is specified in `target_mapping`, then
+> `uri` will be set to `certname`.
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:131]:
```suggestion
> `uri` is set to `certname`.
```

File: documentation/supported_plugins.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+`target_mapping` field:
+
+| Key | Description |
+| --- | ----------- |
+| `certname` | The certname of the node returned from PuppetDB. This is short hand for doing: `facts.trusted.certname`. |
+| `facts.*` | [PQL dot notation](https://puppet.com/docs/puppetdb/latest/api/query/v4/ast.html#dot-notation) facts string such as `facts.os.family` to reference fact value. Dot notation is required for both structured and unstructured facts. |
+
+#### Example usage
+
+Lookup targets with the fact `osfamily: RedHat` and setting:
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:145]:
```suggestion
Look up targets with the fact `osfamily: RedHat` and setting:
```

Comment (hestonhoffman) [inline documentation/supported_plugins.md:145]:
```suggestion
Lookup targets with the fact `osfamily: RedHat` and the following 
configuration values:
```

File: documentation/supported_plugins.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+ * The alias with the fact `hostname`
+ * The name with the fact `certname`
+ * A target fact called `custom_fact` with the `custom_fact` from PuppetDB
+ * A feature from the fact `custom_feature`
+ * The SSH hostname with the fact `networking.interfaces.en0.ipaddress`
+ * The puppetversion var from the fact `puppetversion`
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:151]:
```suggestion
 * The puppetversion variable from the fact `puppetversion`
```

File: documentation/supported_plugins.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+      vars:
+        puppetversion: facts.puppetversion
+```
+
+### `task`
+
+The `task` plugin lets Bolt run a task as a plugin. Tasks will run on the
+`localhost` target without access to any configuration defined in an inventory
+file, but with access to any parameters that are configured. The plugin extracts
+the `value` key from the task output and uses that as the value.
```
Comment (hestonhoffman) [inline documentation/supported_plugins.md:176]:
```suggestion
The `task` plugin lets Bolt run a task as a plugin and extracts
the `value` key from the task output to use as the plugin value. 
Plugin tasks run on the `localhost` target without access to any 
configuration defined in an inventory file, but with access to 
any parameters that you've configured.

For example, you could run a `task` plugin that collects target 
names from a JSON file and interpolates them into a `target` array 
in your inventory file.
```

File: documentation/using_plugins.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (beechtom) [reply documentation/using_plugins.md:6]: I think a link to "Installing modules" would suffice here. Third-party plugins should be written as plugins, so there shouldn't be a need to use the `task` plugin to run them.

File: documentation/using_plugins.md [hunk 14]
```diff
@@ -1,339 +1,213 @@
...
-  - target1.example.com
-config:
+# bolt-defaults.yaml
+inventory-config:
   ssh:
-    user: bolt
-    password:
-      _plugin: env_var
-      var: BOLT_PASSWORD
+    password: bolt
```
[the remark this answers is not in the document]
Comment (beechtom) [reply documentation/using_plugins.md:37]: Removed. I agree that it's a bit confusing.

File: documentation/bolt.ditamap (hunk 3 above)
Comment (beechtom) [reply documentation/bolt.ditamap:19]: Done and done! :)

**Review by hestonhoffman — APPROVED** [review]
This looks good - I've added a small paragraph to Using plugins

**Review by hestonhoffman — APPROVED** [review]

**Pull request merged**